File: android/app/src/main/java/com/yes/yesmusic/methods/spotify/SpotifyDataMappers.java
package com.yes.yesmusic.methods.spotify;

import com.spotify.protocol.types.Album;
import com.spotify.protocol.types.Artist;
import com.spotify.protocol.types.PlayerOptions;
import com.spotify.protocol.types.PlayerRestrictions;
import com.spotify.protocol.types.PlayerState;
import com.spotify.protocol.types.Track;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SpotifyDataMappers {

  private static ArrayList<HashMap<String, Object>> mapArtists(List<Artist> artists) {
    ArrayList<HashMap<String, Object>> artistList = new ArrayList<>();
    for (Artist artist : artists) {
      artistList.add(mapArtist(artist));
    }

    return artistList;
  }

  public static HashMap<String, Object> mapTrack(Track track) {
    HashMap<String, Object> trackMap = new HashMap<>();

    trackMap.put("album", mapAlbum(track.album));
    trackMap.put("artist", mapArtist(track.artist));
    trackMap.put("artists", mapArtists(track.artists));
    trackMap.put("duration", track.duration);
    trackMap.put("imageUri", track.imageUri.raw);
    trackMap.put("isEpisode", track.isEpisode);
    trackMap.put("isPodcast", track.isPodcast);
    trackMap.put("name", track.name);
    trackMap.put("uri", track.uri);

    return trackMap;
  }

  public static HashMap<String, Object> mapAlbum(Album album) {
    HashMap<String, Object> albumMap = new HashMap<>();

    albumMap.put("name", album.name);
    albumMap.put("uri", album.uri);

    return albumMap;
  }

  public static HashMap<String, Object> mapArtist(Artist artist) {
    HashMap<String, Object> artistMap = new HashMap<>();

    artistMap.put("name", artist.name);
    artistMap.put("uri", artist.uri);

    return artistMap;
  }

  public static HashMap<String, Object> mapPlayerState(PlayerState playerState) {
    HashMap<String, Object> stateMap = new HashMap<>();

    stateMap.put("isPaused", playerState.isPaused);
    stateMap.put("playbackOptions", mapPlayerOptions(playerState.playbackOptions));
    stateMap.put("playbackPosition", playerState.playbackPosition);
    stateMap.put("playbackRestrictions", mapPlayerRestrictions(playerState.playbackRestrictions));
    stateMap.put("playbackSpeed", playerState.playbackSpeed);
    stateMap.put("track", mapTrack(playerState.track));

    return stateMap;
  }

  public static HashMap<String, Object> mapPlayerOptions(PlayerOptions playerOptions) {
    HashMap<String, Object> optionsMap = new HashMap<>();

    optionsMap.put("isShuffling", playerOptions.isShuffling);
    optionsMap.put("repeatMode", playerOptions.repeatMode);

    return optionsMap;
  }

  public static HashMap<String, Object> mapPlayerRestrictions(PlayerRestrictions restrictions) {
    HashMap<String, Object> restrictionsMap = new HashMap<>();

    restrictionsMap.put("canRepeatContext", restrictions.canRepeatContext);
    restrictionsMap.put("canRepeatTrack", restrictions.canRepeatTrack);
    restrictionsMap.put("canSeek", restrictions.canSeek);
    restrictionsMap.put("canSkipNext", restrictions.canSkipNext);
    restrictionsMap.put("canSkipPrev", restrictions.canSkipPrev);
    restrictionsMap.put("canToggleShuffle", restrictions.canToggleShuffle);

    return restrictionsMap;
  }
}

File: ios/Runner/SpotifyDataMappers.swift
//
//  SpotifyDataMappers.swift
//  Runner
//
//  Created by Evan Hiroshige on 2/10/19.
//  Copyright © 2019 The Chromium Authors. All rights reserved.
//

import Foundation

class SpotifyDataMappers {
    static func mapArtists(_ artists: Array<SPTAppRemoteArtist>) -> NSArray {
        let artistList = NSMutableArray.init()
        for artist in artists {
            artistList.add(mapArtist(artist))
        }
        return artistList
    }
    
    static func mapTrack(_ track: SPTAppRemoteTrack) -> NSDictionary {
       let trackMap = NSMutableDictionary.init()
        
        trackMap["album"] = mapAlbum(track.album)
        trackMap["artist"] = mapArtist(track.artist)
        trackMap["artists"] = mapArtist(track.artist)
        trackMap["duration"] = track.duration
        
        trackMap["imageUri"] =  track.uri
        trackMap["isEpisode"] =  track.isEpisode
        trackMap["isPodcast"] = track.isPodcast
        trackMap["name"] = track.name
        trackMap["uri"] = track.uri
        
        return trackMap
    }
    
    static func mapAlbum(_ album: SPTAppRemoteAlbum) -> NSDictionary {
        let albumMap = NSMutableDictionary.init()
        
        albumMap["name"] = album.name
        albumMap["uri"] =  album.uri
        
        return albumMap
    }
    
    static func mapArtist(_ artist: SPTAppRemoteArtist) -> NSDictionary{
        let artistMap = NSMutableDictionary.init()
    
        artistMap["name"] = artist.name
        artistMap["uri"] = artist.uri
    
    return artistMap;
    }
    
    static func mapPlayerState(_ playerState: SPTAppRemotePlayerState) -> NSDictionary {
        let stateMap = NSMutableDictionary.init()
    
        stateMap["isPaused"] = playerState.isPaused
    stateMap["playbackOptions"] = mapPlayerOptions(playerState.playbackOptions)
    stateMap["playbackPosition"] = playerState.playbackPosition
        stateMap["playbackRestrictions"] = mapPlayerRestrictions(playerState.playbackRestrictions)
    stateMap["playbackSpeed"] = playerState.playbackSpeed
    stateMap["track"] = mapTrack(playerState.track)
    
    return stateMap
    }
    
    static func mapPlayerOptions(_ playbackOptions: SPTAppRemotePlaybackOptions) -> NSDictionary {
        let optionsMap = NSMutableDictionary.init()
    
    optionsMap["isShuffling"] = playbackOptions.isShuffling
    optionsMap["repeatMode"] = playbackOptions.repeatMode
    
    return optionsMap;
    }
    
    static func mapPlayerRestrictions(_ restrictions: SPTAppRemotePlaybackRestrictions) -> NSDictionary {
    let restrictionsMap = NSMutableDictionary.init()
    
    restrictionsMap["canRepeatContext"] = restrictions.canRepeatContext
    restrictionsMap["canRepeatTrack"] = restrictions.canRepeatTrack
    restrictionsMap["canSeek"] = restrictions.canSeek
    restrictionsMap["canSkipNext"] = restrictions.canSkipNext
    restrictionsMap["canSkipPrev"] = restrictions.canSkipPrevious
    restrictionsMap["canToggleShuffle"] = restrictions.canToggleShuffle
    
    return restrictionsMap
    }
}
